File: common/src/main/java/cc/turtl/chiselmon/system/dashboard/api/endpoints/TimelineHandler.java
package cc.turtl.chiselmon.system.dashboard.api.endpoints;

import cc.turtl.chiselmon.system.dashboard.api.ApiHandler;
import cc.turtl.chiselmon.system.tracker.EncounterDatabase;
import com.sun.net.httpserver.HttpExchange;

import java.io.IOException;
import java.util.List;

public class TimelineHandler extends ApiHandler {

    private record TimelineBucket(long bucket, long count) {
    }

    private record TimelineResponse(String granularity, long bucketMs, List<TimelineBucket> buckets) {
    }

    public TimelineHandler(EncounterDatabase db) {
        super(db);
    }

    @Override
    public void handle(HttpExchange exchange) throws IOException {
        handleRequest(exchange, (timeRange, params) -> {
            // ?granularity=minute|hour (default: hour)
            String granularity = params.getOrDefault("granularity", "hour");
            boolean isMinute = "minute".equalsIgnoreCase(granularity);
            long bucketMs = isMinute ? 60_000L : 3_600_000L;

            List<TimelineBucket> buckets = query("encounters").timeRange(timeRange)
                    .select("FLOOR(encountered_ms / " + bucketMs + ") * " + bucketMs + " AS bucket, COUNT(*) AS cnt")
                    .groupBy("bucket")
                    .orderBy("bucket ASC")
                    .fetchList(rs -> new TimelineBucket(rs.getLong("bucket"), rs.getLong("cnt")));

            return new TimelineResponse(granularity, bucketMs, buckets);
        });
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/system/dashboard/api/endpoints/TimelineHandler.kt
package cc.turtl.chiselmon.system.dashboard.api.endpoints

import cc.turtl.chiselmon.system.dashboard.api.ApiHandler
import cc.turtl.chiselmon.system.tracker.EncounterDatabase
import com.sun.net.httpserver.HttpExchange

class TimelineHandler(db: EncounterDatabase) : ApiHandler(db) {

    private data class TimelineBucket(val bucket: Long, val count: Long)
    private data class TimelineResponse(val granularity: String, val bucketMs: Long, val buckets: List<TimelineBucket>)

    override fun handle(exchange: HttpExchange) {
        handleRequest(exchange) { timeRange, params ->
            // ?granularity=minute|hour (default: hour)
            val granularity = params["granularity"] ?: "hour"
            val isMinute = granularity.equals("minute", ignoreCase = true)
            val bucketMs = if (isMinute) 60_000L else 3_600_000L

            val buckets = query("encounters").timeRange(timeRange)
                .select("FLOOR(encountered_ms / $bucketMs) * $bucketMs AS bucket, COUNT(*) AS cnt")
                .groupBy("bucket")
                .orderBy("bucket ASC")
                .fetchList { rs -> TimelineBucket(rs.getLong("bucket"), rs.getLong("cnt")) }

            TimelineResponse(granularity, bucketMs, buckets)
        }
    }
}
